crates/nu-system/src/foreground.rs

```diff
@@ -0,0 +1,117 @@
+        })
+    }
+}
+
+impl AsMut<Child> for ForegroundChild {
+    fn as_mut(&mut self) -> &mut Child {
+        &mut self.inner
+    }
+}
+
+impl Drop for ForegroundChild {
+    fn drop(&mut self) {
+        // It's ok to use here because we have called `set_foreground` during creation.
+        unsafe { fg_process_setup::reset_foreground_id() }
+    }
+}
+
+// It's a simpler version of fish shell's external process handling.
+#[cfg(target_family = "unix")]
+mod fg_process_setup {
+    use crate::signal::{block, unblock};
+    use nix::unistd::{self, Pid};
+    use std::os::unix::prelude::CommandExt;
+
+    pub(super) fn prepare_to_foreground(external_command: &mut std::process::Command) {
+        unsafe {
+            block();
+            // Safety:
+            // POSIX only allows async-signal-safe functions to be called.
+            // And `setpgid` is async-signal-safe function according to:
+            // https://manpages.ubuntu.com/manpages/bionic/man7/signal-safety.7.html
+            // So we're ok to invoke `libc::setpgid` inside `pre_exec`.
+            external_command.pre_exec(|| {
+                // make the command startup with new process group.
+                // The process group id must be the same as external commands' pid.
+                // Or else we'll failed to set it as foreground process.
+                // For more information, check `fork_child_for_process` function:
+                // https://github.com/fish-shell/fish-shell/blob/023042098396aa450d2c4ea1eb9341312de23126/src/exec.cpp#L398
+                if let Err(e) = unistd::setpgid(Pid::from_raw(0), Pid::from_raw(0)) {
+                    println!("ERROR: setpgid for external failed, result: {e:?}");
+                }
+                Ok(())
+            });
+        }
+    }
+
+    // If `prepare_to_foreground` function is not called, the function will fail with silence and do nothing.
+    pub(super) fn set_foreground(process: &std::process::Child) {
+        if atty::is(atty::Stream::Stdin) {
+            if let Err(e) =
+                nix::unistd::tcsetpgrp(nix::libc::STDIN_FILENO, Pid::from_raw(process.id() as i32))
+            {
+                println!("ERROR: set foreground id failed, tcsetpgrp result: {e:?}");
+            }
+        }
+    }
+
+    /// Reset foreground to current process, unblock `SIGTSTP`, `SIGTTOU`, `SIGTTIN`, `SIGCHLD`
+    ///
+    /// ## Safety
+    /// It can only be called when you have called `set_foreground`, or results in undefined behavior.
+    pub(super) unsafe fn reset_foreground_id() {
+        if atty::is(atty::Stream::Stdin) {
+            if let Err(e) = nix::unistd::tcsetpgrp(nix::libc::STDIN_FILENO, unistd::getpgrp()) {
+                println!("ERROR: reset foreground id failed, tcsetpgrp result: {e:?}");
+            }
+        }
+        unblock()
+    }
+}
+
+// TODO: investigate if we can set foreground process through windows system call.
+#[cfg(target_family = "windows")]
```

Comment:
The blank impl probably should be `cfg(not(unix))` rather than windows specifically…